File: mobile/src/main/java/info/papdt/express/helper/ui/adapter/CompanyListAdapter.java
package info.papdt.express.helper.ui.adapter;

import android.graphics.drawable.ColorDrawable;
import android.support.v7.widget.AppCompatTextView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.ArrayList;

import de.hdodenhof.circleimageview.CircleImageView;
import info.papdt.express.helper.R;
import info.papdt.express.helper.api.PackageApi;
import info.papdt.express.helper.support.ColorGenerator;
import info.papdt.express.helper.ui.common.SimpleRecyclerViewAdapter;

public class CompanyListAdapter extends SimpleRecyclerViewAdapter {

	private ArrayList<PackageApi.CompanyInfo.Company> list;

	public CompanyListAdapter(RecyclerView recyclerView, ArrayList<PackageApi.CompanyInfo.Company> list) {
		super(recyclerView);
		this.list = list;
	}

	public void setList(ArrayList<PackageApi.CompanyInfo.Company> list) {
		this.list = list;
	}

	@Override
	public ClickableViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
		bindContext(parent.getContext());
		return new ItemHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_list_company, parent, false));
	}

	@Override
	public void onBindViewHolder(ClickableViewHolder holder, int pos) {
		super.onBindViewHolder(holder, pos);
		if (holder instanceof ItemHolder) {
			ItemHolder itemHolder = (ItemHolder) holder;
			itemHolder.titleText.setText(getItem(pos).getName());
			itemHolder.otherText.setText(getItem(pos).getPhone() != null ? getItem(pos).getPhone() : getItem(pos).getWebsite());
			itemHolder.otherText.setVisibility(itemHolder.otherText.getText() != null ? View.VISIBLE : View.INVISIBLE);

			/** Set up the logo */
			itemHolder.logoView.setImageDrawable(new ColorDrawable(ColorGenerator.MATERIAL.getColor(getItem(pos).getName())));
			itemHolder.firstCharText.setText(getItem(pos).getName().substring(0, 1));
		}
	}

	@Override
	public int getItemCount() {
		return list == null ? 0 : list.size();
	}

	public PackageApi.CompanyInfo.Company getItem(int pos) {
		return list.get(pos);
	}

	public class ItemHolder extends ClickableViewHolder {

		AppCompatTextView titleText, otherText;
		CircleImageView logoView;
		TextView firstCharText;

		public ItemHolder(View itemView) {
			super(itemView);
			titleText = (AppCompatTextView) itemView.findViewById(R.id.tv_title);
			otherText = (AppCompatTextView) itemView.findViewById(R.id.tv_other);
			logoView = (CircleImageView) itemView.findViewById(R.id.iv_logo);
			firstCharText = (TextView) itemView.findViewById(R.id.tv_first_char);
		}

	}

}

File: mobile/src/main/kotlin/info/papdt/express/helper/ui/adapter/CompanyListAdapter.kt
package info.papdt.express.helper.ui.adapter

import android.graphics.drawable.ColorDrawable
import androidx.appcompat.widget.AppCompatTextView
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import java.util.ArrayList

import de.hdodenhof.circleimageview.CircleImageView
import info.papdt.express.helper.R
import info.papdt.express.helper.api.Kuaidi100PackageApi
import info.papdt.express.helper.ui.common.SimpleRecyclerViewAdapter

class CompanyListAdapter(
        recyclerView: RecyclerView,
        private var list: ArrayList<Kuaidi100PackageApi.CompanyInfo.Company>?
) : SimpleRecyclerViewAdapter(recyclerView) {

    fun setList(list: ArrayList<Kuaidi100PackageApi.CompanyInfo.Company>) {
        this.list = list
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SimpleRecyclerViewAdapter.ClickableViewHolder {
        bindContext(parent.context)
        return ItemHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_list_company, parent, false))
    }

    override fun onBindViewHolder(holder: SimpleRecyclerViewAdapter.ClickableViewHolder, position: Int) {
        super.onBindViewHolder(holder, position)
        val item = getItem(position)
        if (holder is ItemHolder) {
            holder.titleText.text = item.name
            holder.otherText.text = if (item.phone != null) item.phone else item.website
            holder.otherText.visibility = if (holder.otherText.text != null) View.VISIBLE else View.INVISIBLE

            /** Set up the logo  */
            holder.firstCharText.text = item.name.substring(0, 1)
            item.getPalette().let {
                holder.logoView.setImageDrawable(
                        ColorDrawable(it.getPackageIconBackground(context!!)))
                holder.firstCharText.setTextColor(
                        it.getPackageIconForeground(context!!))
            }
        }
    }

    override fun getItemCount(): Int = list?.size ?: 0

    fun getItem(pos: Int): Kuaidi100PackageApi.CompanyInfo.Company = list!![pos]

    inner class ItemHolder(itemView: View) : SimpleRecyclerViewAdapter.ClickableViewHolder(itemView) {

        internal var titleText: AppCompatTextView = itemView.findViewById(R.id.tv_title)
        internal var otherText: AppCompatTextView = itemView.findViewById(R.id.tv_other)
        internal var logoView: CircleImageView = itemView.findViewById(R.id.iv_logo)
        internal var firstCharText: TextView = itemView.findViewById(R.id.tv_first_char)

    }

}
